src/Npgsql/NpgsqlRecordDataReader.cs

```diff
@@ -0,0 +1,327 @@
+            CheckResultSet();
+
+            var count = Math.Min(FieldCount, values.Length);
+            for (var i = 0; i < count; i++)
+                values[i] = GetValue(i);
+            return count;
+        }
+
+        public override bool IsDBNull(int ordinal)
+            => CheckRowAndColumnAndSeek(ordinal).Length == -1;
+
+        public override T GetFieldValue<T>(int ordinal)
+        {
+            if (typeof(T) == typeof(Stream))
+                return (T)(object)GetStream(ordinal);
+
+            if (typeof(T) == typeof(TextReader))
+                return (T)(object)GetTextReader(ordinal);
+
+            var field = CheckRowAndColumnAndSeek(ordinal);
+
+            if (field.Length == -1)
+            {
+                // When T is a Nullable<T> (and only in that case), we support returning null
+                if (NullableHandler<T>.Exists)
+                    return default!;
+
+                if (typeof(T) == typeof(object))
+                    return (T)(object)DBNull.Value;
+
+                throw new InvalidCastException("field is null");
+            }
+
+            var position = Buffer.ReadPosition;
+            return NullableHandler<T>.Exists
+                ? NullableHandler<T>.Read(field.Handler, Buffer, field.Length, fieldDescription: null)
+                : typeof(T) == typeof(object)
+                    ? (T)field.Handler.ReadAsObject(Buffer, field.Length, fieldDescription: null)
+                    : field.Handler.Read<T>(Buffer, field.Length, fieldDescription: null);
+        }
+
+        public override Type GetProviderSpecificFieldType(int ordinal)
+        {
+            var column = CheckRowAndColumn(ordinal);
+            return TypeMapper.GetByOID(column.Oid).GetProviderSpecificFieldType();
+        }
+
+        public override object GetProviderSpecificValue(int ordinal)
+        {
+            var column = CheckRowAndColumnAndSeek(ordinal);
+            if (column.Length == -1)
+                return DBNull.Value;
+            return column.Handler.ReadPsvAsObject(Buffer, column.Length);
+        }
+
+        public override int GetProviderSpecificValues(object[] values)
+        {
+            if (values == null)
+                throw new ArgumentNullException(nameof(values));
+            CheckResultSet();
+
+            var count = Math.Min(FieldCount, values.Length);
+            for (var i = 0; i < count; i++)
+                values[i] = GetProviderSpecificValue(i);
+            return count;
+        }
+
+        public override bool Read()
+        {
+            CheckResultSet();
+
+            Buffer.ReadPosition = _nextRowBufferPos;
+            if (_nextRowIndex >= _numRows)
+                return false;
+
+            if (_nextRowIndex++ != 0)
+                Buffer.ReadInt32(); // Length of record
+
+            var numColumns = Buffer.ReadInt32();
+
+            if (_columns == null || _columns.Length != numColumns)
```

Comment:
We could optimize a bit for the heterogeneous row scenario and use a List, so that the size can change with reallocating the whole array for each row.